PoC-updater/updater.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2021 Red Hat, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/RedHatInsights/ccx-notification-service/conf"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"updater/types"
+)
+
+// Configuration-related constants
+const (
+	configFileEnvVariableName = "NOTIFICATION_DIFFER_CONFIG_FILE"
+	defaultConfigFileName     = "config"
+)
+
+// Messages
+const (
+	operationFailedMessage = "Operation failed"
+)
+
+// Exit codes
+const (
+	// ExitStatusOK means that the tool finished with success
+	ExitStatusOK = iota
+	// ExitStatusError is a general error code
+	ExitStatusError
+	// ExitStatusStorageError is returned in case of any consumer-related error
+	ExitStatusStorageError
+)
+
+func unmarshalReport(reportAsString types.ClusterReport) (types.Report, error) {
+	var report types.Report
+	err := json.Unmarshal([]byte(reportAsString), &report)
+	if err != nil {
+		log.Err(err).Msg("Unmarshal")
+		return report, err
+	}
+	return report, err
+}
+
+func process(storage Storage) {
+	const testedOrgID = 1
+	const testedCluster = "11111111-1111-1111-1111-111111111111"
+
+	// read newest report for a cluster
+	rNewAsString, timestamp, err := storage.ReadReportForCluster(testedOrgID, testedCluster)
+	if err != nil {
+		log.Error().Err(err).Msg("Read report for cluster")
+		return
+	}
+	log.Info().Time("time", time.Time(timestamp)).Msg("Read")
+
+	c := types.ClusterEntry{
+		OrgID:       testedOrgID,
+		ClusterName: testedCluster,
+	}
+
+	// read older reports for a cluster
+	reported, err := storage.ReadLastNNotificationRecords(c, 10)
+	if err != nil {
+		log.Error().Err(err).Msg("Read last n reports")
+	}
+	for _, r := range reported {
+		rOldAsString := r.Report
+		log.Info().Time("notified at", time.Time(r.NotifiedAt)).Msg("notified record")
+
+		notify, err := CompareReports(rOldAsString, rNewAsString)
+		if err != nil {
+			log.Error().Err(err).Msg("Unable to compare")
+		}
+		log.Info().Bool("resolution", notify).Msg("Should notify user")
+		fmt.Println()
+	}
+}
+
+// Function issuesEqual compares two issues from reports
+func issuesEqual(issue1 types.ReportItem, issue2 types.ReportItem) bool {
+	if issue1.Type == issue2.Type &&
+		issue1.Module == issue2.Module &&
+		issue1.ErrorKey == issue2.ErrorKey &&
+		bytes.Compare(issue1.Details, issue2.Details) == 0 {
```

Comment:
could use bytes.Equal(issue1.Details, issue2.Details).
[Docs](https://golang.org/pkg/bytes/#Compare) say to prefer Equal over Compare for equality comparison

Reply:
ah it already exist? cool I was waiting for this function from Go 1.10 or so :)

Reply:
fixed :)